src/python/pants/engine/legacy/structs.py

```diff
@@ -38,7 +39,18 @@ def get_sources(self):
     refactor how deferred sources are implemented.
       see: https://github.com/pantsbuild/pants/issues/2997
     """
+    source = getattr(self, 'source', None)
     sources = getattr(self, 'sources', None)
+
+    if source and sources:
+      raise Target.IllegalArgument(
+        self.address.spec,
+        'Cannot specify both source and sources attribute.'
+      )
+
+    if source:
```

Comment:
This should probably be `if source is not None`, to minimize edgecases around empty strings.

Reply:
Done